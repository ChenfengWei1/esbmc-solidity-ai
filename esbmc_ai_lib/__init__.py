version: str = "0.01"
