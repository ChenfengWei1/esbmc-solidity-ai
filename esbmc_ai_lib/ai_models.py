# Author: Yiannis Charalambous

from typing import NamedTuple
from tiktoken import get_encoding, encoding_for_model

from .chat_api.chat_api import ChatAPI


class AIModel(NamedTuple):
    name: str = ""
    tokens: int = 0


AI_MODEL_GPT3 = AIModel("gpt-3.5-turbo", 4096)
AI_MODEL_GPT3_16K = AIModel("gpt-3.5-turbo-16k", 16384)
AI_MODEL_GPT4 = AIModel("gpt-4", 8192)
AI_MODEL_GPT4_32k = AIModel("gpt-4-32k", 32768)

# TODO Make it dynamically set from config.
AI_MODEL_TEXTGEN = AIModel("text-generation-inference", 8192)

models: list[AIModel] = [
    AI_MODEL_GPT3,
    AI_MODEL_GPT3_16K,
    AI_MODEL_GPT4,
    AI_MODEL_GPT4_32k,
    AI_MODEL_TEXTGEN,
]


def is_valid_ai_model(name: str) -> bool:
    for model in models:
        if model.name == name:
            return True
    return False


def num_tokens_from_messages(messages, model: AIModel):
    """Returns the number of tokens used by a list of messages.
    Source: https://platform.openai.com/docs/guides/chat/introduction"""
    model_name: str = model.name
    try:
        encoding = encoding_for_model(model_name)
    except KeyError:
        encoding = get_encoding("cl100k_base")

    # note: future models may deviate from this
    if is_valid_ai_model(model.name):
        num_tokens = 0
        for message in messages:
            # every message follows <im_start>{role/name}\n{content}<im_end>\n
            num_tokens += 4
            for key, value in message.items():
                num_tokens += len(encoding.encode(value))
                # if there's a name, the role is omitted
                if key == "name":
                    # role is always required and always 1 token
                    num_tokens += -1
        # every reply is primed with <im_start>assistant
        num_tokens += 2
        return num_tokens
    else:
        # See https://github.com/openai/openai-python/blob/main/chatml.md for
        # information on how messages are converted to tokens.
        raise NotImplementedError(
            f"num_tokens_from_messages() is not presently implemented for model {model}."
        )
