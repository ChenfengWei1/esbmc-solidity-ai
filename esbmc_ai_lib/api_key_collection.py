# Author: Yiannis Charalambous


from typing import NamedTuple


class APIKeyCollection(NamedTuple):
    openai: str = ""
    huggingface: str = ""
