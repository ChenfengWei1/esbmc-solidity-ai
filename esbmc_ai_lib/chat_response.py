# Author: Yiannis Charalambous

from enum import Enum
from typing import NamedTuple

from langchain.schema import AIMessage, BaseMessage, HumanMessage, SystemMessage


class FinishReason(Enum):
    # API response still in progress or incomplete
    null = 0
    # API returned complete model output
    stop = 1
    # Incomplete model output due to max_tokens parameter or token limit
    length = 2
    # Omitted content due to a flag from our content filters
    content_filter = 3


class ChatResponse(NamedTuple):
    message: BaseMessage
    total_tokens: int = 0
    finish_reason: FinishReason = FinishReason.null


def json_to_base_message(json_string: dict) -> BaseMessage:
    """Converts a json representation of messages (such as in config.json),
    into LangChain object messages. The three recognized roles are:
    1. System
    2. AI
    3. Human"""
    role: str = json_string["role"]
    content: str = json_string["content"]
    if role == "System":
        return SystemMessage(content=content)
    elif role == "AI":
        return AIMessage(content=content)
    elif role == "Human":
        return HumanMessage(content=content)
    else:
        raise Exception()
