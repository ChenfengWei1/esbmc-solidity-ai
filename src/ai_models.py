AI_MODEL_GPT3 = "gpt-3.5-turbo"
AI_MODEL_GPT4 = "gpt-4"


def is_valid_ai_model(name: str) -> bool:
    return name == AI_MODEL_GPT3 or name == AI_MODEL_GPT4
