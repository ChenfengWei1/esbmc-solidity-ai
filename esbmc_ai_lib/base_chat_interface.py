# Author: Yiannis Charalambous

from abc import abstractmethod
from typing import NamedTuple
import openai
from tiktoken import get_encoding, encoding_for_model

from .ai_models import AI_MODEL_GPT3, AI_MODEL_GPT4

MAX_TOKENS_GPT3TURBO: int = 4096


def num_tokens_from_messages(messages, model=AI_MODEL_GPT3):
    """Returns the number of tokens used by a list of messages.
    Source: https://platform.openai.com/docs/guides/chat/introduction"""
    try:
        encoding = encoding_for_model(model)
    except KeyError:
        encoding = get_encoding("cl100k_base")
    # note: future models may deviate from this
    if model.startswith(AI_MODEL_GPT3) or model.startswith(AI_MODEL_GPT4):
        num_tokens = 0
        for message in messages:
            # every message follows <im_start>{role/name}\n{content}<im_end>\n
            num_tokens += 4
            for key, value in message.items():
                num_tokens += len(encoding.encode(value))
                # if there's a name, the role is omitted
                if key == "name":
                    # role is always required and always 1 token
                    num_tokens += -1
        # every reply is primed with <im_start>assistant
        num_tokens += 2
        return num_tokens
    else:
        # See https://github.com/openai/openai-python/blob/main/chatml.md for
        # information on how messages are converted to tokens.
        raise NotImplementedError(
            f"num_tokens_from_messages() is not presently implemented for model {model}."
        )


# API returned complete model output
FINISH_REASON_STOP: str = "stop"
# Incomplete model output due to max_tokens parameter or token limit
FINISH_REASON_LENGTH: str = "length"
# Omitted content due to a flag from our content filters
FINISH_REASON_CONTENT_FILTER: str = "content_filter"
# API response still in progress or incomplete
FINISH_REASON_NULL: str = "null"


class ChatResponse(NamedTuple):
    base_message: object = None
    finish_reason: str = FINISH_REASON_NULL
    role: str = ""
    message: str = ""
    total_tokens: int = 0


class BaseChatInterface(object):
    max_tokens = MAX_TOKENS_GPT3TURBO

    protected_messages: list
    messages: list
    model_name: str
    temperature: float

    def __init__(
        self,
        system_messages: list,
        model: str = "gpt-3.5-turbo",
        temperature: float = 1.0,
    ) -> None:
        super().__init__()
        self.model_name = model
        self.temperature = temperature

        self.protected_messages = system_messages.copy()
        self.messages = system_messages.copy()

    @abstractmethod
    def compress_message_stack(self) -> None:
        raise NotImplementedError()

    def push_to_message_stack(
        self, role: str, message: str, protected: bool = False
    ) -> None:
        message_struct = {"role": role, "content": message}
        if protected:
            self.protected_messages.append(message_struct)
        self.messages.append(message_struct)

    # Returns an OpenAI object back.
    def send_message(self, message: str, protected: bool = False) -> ChatResponse:
        """Sends a message to the AI model. Returns solution. If the message
        stack fills up, the command will exit with no changes to the message
        stack."""
        # See if the new stack if over the limit.
        new_stack = [*self.messages, {"role": "user", "content": message}]

        # Check if message is too long and exit.
        msg_tokens: int = num_tokens_from_messages(new_stack, self.model_name)
        if msg_tokens > self.max_tokens:
            response: ChatResponse = ChatResponse(
                finish_reason=FINISH_REASON_LENGTH,
            )
            return response

        completion = openai.ChatCompletion.create(
            model=self.model_name,
            messages=new_stack,
            temperature=self.temperature,
        )

        response = ChatResponse(
            base_message=completion,
            role=completion.choices[0].message.role,
            message=completion.choices[0].message.content,
            finish_reason=completion.choices[0].finish_reason,
            total_tokens=completion.usage.total_tokens,
        )

        # If the response is OK then add to stack.
        if response.finish_reason == FINISH_REASON_STOP:
            self.push_to_message_stack("user", message, protected)
            self.push_to_message_stack(response.role, response.message, protected)

        return response
