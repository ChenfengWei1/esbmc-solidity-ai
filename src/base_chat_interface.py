# Author: Yiannis Charalambous

import openai
from tiktoken import get_encoding, encoding_for_model

from src.ai_models import AI_MODEL_GPT3, AI_MODEL_GPT4

MAX_TOKENS_GPT3TURBO: int = 4096


def num_tokens_from_messages(messages, model=AI_MODEL_GPT3):
    """Returns the number of tokens used by a list of messages.
    Source: https://platform.openai.com/docs/guides/chat/introduction"""
    try:
        encoding = encoding_for_model(model)
    except KeyError:
        encoding = get_encoding("cl100k_base")
    # note: future models may deviate from this
    if model.startswith(AI_MODEL_GPT3) or model.startswith(AI_MODEL_GPT4):
        num_tokens = 0
        for message in messages:
            # every message follows <im_start>{role/name}\n{content}<im_end>\n
            num_tokens += 4
            for key, value in message.items():
                num_tokens += len(encoding.encode(value))
                # if there's a name, the role is omitted
                if key == "name":
                    # role is always required and always 1 token
                    num_tokens += -1
        # every reply is primed with <im_start>assistant
        num_tokens += 2
        return num_tokens
    else:
        # See https://github.com/openai/openai-python/blob/main/chatml.md for
        # information on how messages are converted to tokens.
        raise NotImplementedError(
            f"num_tokens_from_messages() is not presently implemented for model {model}."
        )


class BaseChatInterface(object):
    max_tokens = MAX_TOKENS_GPT3TURBO

    system_messages: list
    messages: list
    model_name: str
    temperature: float

    def __init__(
        self,
        system_messages: list,
        model: str = "gpt-3.5-turbo",
        temperature: float = 1.0,
    ) -> None:
        super().__init__()
        self.system_messages = system_messages
        self.messages = self.system_messages
        self.model_name = model
        self.temperature = temperature

    def push_to_message_stack(self, role: str, message: str) -> None:
        self.messages.append({"role": role, "content": message})

    # Returns an OpenAI object back.
    def send_message(self, message: str):
        self.push_to_message_stack("user", message)

        # Check if necessary to shorten.
        msg_tokens: int = num_tokens_from_messages(self.messages, self.model_name)
        if msg_tokens > self.max_tokens:
            print(
                f"Max tokens reached... Conversation too long {msg_tokens}/{self.max_tokens}. Exiting..."
            )
            exit(2)

        completion = openai.ChatCompletion.create(
            model=self.model_name,
            messages=self.messages,
            temperature=self.temperature,
        )

        response_role = completion.choices[0].message.role
        response_message = completion.choices[0].message.content
        self.push_to_message_stack(response_role, response_message)

        return completion
