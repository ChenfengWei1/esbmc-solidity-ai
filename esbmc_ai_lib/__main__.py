#!/usr/bin/env python3

# Author: Yiannis Charalambous 2023

import os

# Enables arrow key functionality for input(). Do not remove import.
import readline

import argparse
import openai

import esbmc_ai_lib.config as config

from esbmc_ai_lib.commands import (
    ChatCommand,
    FixCodeCommand,
    HelpCommand,
    ExitCommand,
    VerifyCodeCommand,
)

from esbmc_ai_lib.loading_widget import LoadingWidget
from esbmc_ai_lib.user_chat import ChatInterface
from esbmc_ai_lib.base_chat_interface import (
    ChatResponse,
    FINISH_REASON_LENGTH,
    FINISH_REASON_STOP,
)
from esbmc_ai_lib.logging import printv
from esbmc_ai_lib.esbmc_util import esbmc


commands: list[ChatCommand] = []
help_command: HelpCommand = HelpCommand()
fix_code_command: FixCodeCommand = FixCodeCommand()
verify_code_command: VerifyCodeCommand = VerifyCodeCommand()
exit_command: ExitCommand = ExitCommand()

chat: ChatInterface

HELP_MESSAGE: str = """Tool that passes ESBMC output into ChatGPT and allows for natural language
explanations. Type /help in order to view available commands."""

ESBMC_HELP: str = """Additional ESBMC Arguments - The following are useful
arguments that can be added after the filename to modify ESBMC functionality.
For all the options, run ESBMC with -h as a parameter:

  --no-assertions                  ignore assertions
  --no-bounds-check                do not do array bounds check
  --no-div-by-zero-check           do not do division by zero check
  --no-pointer-check               do not do pointer check
  --no-align-check                 do not check pointer alignment
  --no-pointer-relation-check      do not check pointer relations
  --no-unlimited-scanf-check       do not do overflow check for scanf/fscanf
                                   with unlimited character width.
  --nan-check                      check floating-point for NaN
  --memory-leak-check              enable memory leak check
  --overflow-check                 enable arithmetic over- and underflow check
  --ub-shift-check                 enable undefined behaviour check on shift
                                   operations
  --struct-fields-check            enable over-sized read checks for struct
                                   fields
  --deadlock-check                 enable global and local deadlock check with
                                   mutex
  --data-races-check               enable data races check
  --lock-order-check               enable for lock acquisition ordering check
  --atomicity-check                enable atomicity check at visible
                                   assignments
  --stack-limit bits (=-1)         check if stack limit is respected
  --error-label label              check if label is unreachable
  --force-malloc-success           do not check for malloc/new failure
"""


def init_check_health(verbose: bool) -> None:
    def printv(m) -> None:
        if verbose:
            print(m)

    printv("Performing init health check...")
    # Check that the .env file exists.
    if os.path.exists(".env"):
        printv("Environment file has been located")
    else:
        print("Error: .env file is not found in project directory")
        exit(3)


def check_health() -> None:
    printv("Performing health check...")
    # Check that ESBMC exists.
    if os.path.exists(config.esbmc_path):
        printv("ESBMC has been located")
    else:
        print(f"Error: ESBMC could not be found in {config.esbmc_path}")
        exit(3)


def get_src(path: str) -> str:
    with open(path, mode="r") as file:
        content = file.read()
        return str(content)


def print_assistant_response(
    chat: ChatInterface,
    response: ChatResponse,
    raw_responses: bool = False,
    hide_stats: bool = False,
) -> None:
    if raw_responses:
        print(response.base_message)
        return

    print(f"{response.role}: {response.message}\n\n")

    if not hide_stats:
        print(
            "Stats:",
            f"total tokens: {response.total_tokens},",
            f"max tokens: {chat.max_tokens}",
            f"finish reason: {response.finish_reason}",
        )


def init_commands() -> None:
    """Function that handles initializing commands. Each command needs to be added
    into the commands array in order for the command to register to be called by
    the user and also register in the help system."""
    # Bus Signals:
    # fix_code_command.on_solution_signal.add_listener(chat.set_solution)
    # fix_code_command.on_solution_signal.add_listener(verify_code_command.set_solution)

    # Setup Help command and commands list.
    global help_command
    commands.extend(
        [
            help_command,
            exit_command,
            fix_code_command,
            verify_code_command,
        ]
    )
    help_command.set_commands(commands)


def build_system_messages(source_code: str, esbmc_output: str) -> list:
    """Build the setup messages from either the provided default settings or from
    the loaded files."""
    printv("Loading system messages")
    system_messages: list = []
    if len(config.chat_prompt_user_mode.system_messages) > 0:
        system_messages.extend(config.chat_prompt_user_mode.system_messages)
    else:
        raise RuntimeError("Chat mode system messages could not be loaded from config.")

    # Add the introduction of code prompts. TODO Make these loaded from config
    # too in the future.
    system_messages.extend(
        [
            {
                "role": "system",
                "content": f"Reply OK if you understand that the following text is the program source code:\n\n{source_code}",
            },
            {"role": "assistant", "content": "OK"},
            {
                "role": "system",
                "content": f"Reply OK if you understand that the following text is the output from ESBMC:\n\n{esbmc_output}",
            },
            {"role": "assistant", "content": "OK"},
        ]
    )

    return system_messages


def main() -> None:
    parser = argparse.ArgumentParser(
        prog="ESBMC-ChatGPT",
        description=HELP_MESSAGE,
        # argparse.RawDescriptionHelpFormatter allows the ESBMC_HELP to display
        # properly.
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=f"Made by Yiannis Charalambous\n\n{ESBMC_HELP}",
    )

    parser.add_argument("filename", help="The filename to pass to esbmc.")
    parser.add_argument(
        "remaining",
        nargs=argparse.REMAINDER,
        help="Any arguments after the filename will be passed to ESBMC as parameters.",
    )

    parser.add_argument(
        "-v",
        "--verbose",
        action="store_true",
        default=False,
        help="Output will be verbose.",
    )

    parser.add_argument(
        "-r",
        "--raw-output",
        action="store_true",
        default=False,
        help="Responses from AI model will be shown raw.",
    )

    parser.add_argument(
        "-m",
        "--ai-model",
        default="",
        help="Which AI model to use.",
    )

    parser.add_argument(
        "-a",
        "--append",
        action="store_true",
        default=False,
        help="Any ESBMC parameters passed after the file name will be appended to the ones set in the config file, or the default ones if config file options are not set.",
    )

    args = parser.parse_args()

    print("ESBMC-AI")
    print("Made by Yiannis Charalambous")
    print()

    init_check_health(args.verbose)

    config.load_envs()
    config.load_config(config.cfg_path)
    config.load_args(args)

    check_health()

    anim: LoadingWidget = LoadingWidget()

    # Read the source code and esbmc output.
    printv("Reading source code...")
    print(f"Running ESBMC with {config.esbmc_params}\n")
    source_code: str = get_src(args.filename)

    anim.start("ESBMC is processing... Please Wait")
    exit_code, esbmc_output = esbmc(
        path=args.filename,
        esbmc_params=config.esbmc_params,
    )
    anim.stop()

    # ESBMC will output 0 for verification success and 1 for verification
    # failed, if anything else gets thrown, it's an ESBMC error.
    if exit_code == 0:
        print("Success!")
        print(esbmc_output)
        exit(0)
    elif exit_code != 1:
        print(f"ESBMC exit code: {exit_code}")
        exit(1)

    # Prepare the chat.
    printv("Initializing OpenAI")
    openai.api_key = config.openai_api_key

    # Inject output for ai.
    system_messages: list = build_system_messages(source_code, esbmc_output)

    global chat
    chat = ChatInterface(
        system_messages=system_messages,
        model=config.ai_model,
        temperature=config.chat_temperature,
    )
    printv(f"Using AI Model: {chat.model_name}")

    # Show the initial output.
    response: ChatResponse
    if len(config.chat_prompt_user_mode.initial_prompt) > 0:
        printv("Using initial prompt from file...\n")
        anim.start("Model is parsing ESBMC output... Please Wait")
        response = chat.send_message(config.chat_prompt_user_mode.initial_prompt)
        anim.stop()
    else:
        raise RuntimeError("User mode initial prompt not found in config.")

    print_assistant_response(chat, response, config.raw_responses)
    print(
        "ESBMC-AI: Type '/help' to view the available in-chat commands, along",
        "with useful prompts to ask the AI model...",
    )

    printv("Initializing commands...")
    init_commands()

    while True:
        # Get user input.
        user_message = input(">: ")

        # Check if it is a command, if not, then pass it to the chat interface.
        if user_message.startswith("/"):
            command: str = user_message[1:]
            if command == fix_code_command.command_name:
                print()
                print("ESBMC-AI will generate a fix for the code...")

                error, solution = fix_code_command.execute(
                    source_code=source_code,
                    esbmc_output=esbmc_output,
                )

                if not error:
                    # Let the AI model know about the corrected code.
                    printv("Informing Chat AI about correct code...")
                    chat.set_solution(solution)
                continue
            else:
                # Commands without parameters or returns are handled automatically.
                found: bool = False
                for cmd in commands:
                    if cmd.command_name == command:
                        found = True
                        cmd.execute()
                        break

                if not found:
                    print("Error: Unknown command...")
                continue
        elif user_message == "":
            continue
        else:
            print()

        while True:
            # Send user message to AI model and process.
            anim.start("Generating response... Please Wait")
            response = chat.send_message(user_message)
            anim.stop()
            if response.finish_reason == FINISH_REASON_STOP:
                break
            elif response.finish_reason == FINISH_REASON_LENGTH:
                anim.start(
                    "Message stack limit reached. Shortening message stack... Please Wait"
                )
                chat.compress_message_stack()
                anim.stop()
                continue
            else:
                raise NotImplementedError(
                    f"User Chat Mode: Finish Reason: {response.finish_reason}"
                )

        print_assistant_response(
            chat,
            response,
            config.raw_responses,
        )


if __name__ == "__main__":
    main()
