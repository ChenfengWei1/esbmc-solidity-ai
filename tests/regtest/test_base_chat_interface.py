# Author: Yiannis Charalambous


from langchain.llms.fake import FakeListLLM
from langchain.schema import BaseMessage, HumanMessage, AIMessage, SystemMessage

from esbmc_ai_lib.ai_models import AIModel
from esbmc_ai_lib.chat_response import ChatResponse
from esbmc_ai_lib.base_chat_interface import BaseChatInterface


def test_push_message_stack(regtest) -> None:
    llm: FakeListLLM = FakeListLLM(responses=[])

    ai_model: AIModel = AIModel("test", 1024)

    chat: BaseChatInterface = BaseChatInterface(
        ai_model=ai_model,
        system_messages=[],
        llm=llm,
    )

    messages: list[BaseMessage] = [
        AIMessage(content="Test 1"),
        HumanMessage(content="Test 2"),
        SystemMessage(content="Test 3"),
    ]

    chat.push_to_message_stack(message=messages[0], protected=False)
    chat.push_to_message_stack(message=messages[1], protected=True)
    chat.push_to_message_stack(message=messages[2], protected=True)

    with regtest:
        print(chat.protected_messages)
        print(chat.messages)


def test_send_message(regtest) -> None:
    responses: list[str] = ["OK 1", "OK 2", "OK 3"]
    llm: FakeListLLM = FakeListLLM(responses=responses)

    ai_model: AIModel = AIModel("test", 6)

    sys: list[BaseMessage] = []
    chat: BaseChatInterface = BaseChatInterface(
        ai_model=ai_model,
        system_messages=sys,
        llm=llm,
    )

    chat_responses: list[ChatResponse] = [
        chat.send_message("Test 1"),
        chat.send_message("Test 2"),
        chat.send_message("Test 3"),
    ]

    with regtest:
        print(chat_responses)
        print(chat.protected_messages)
        print(chat.messages)
