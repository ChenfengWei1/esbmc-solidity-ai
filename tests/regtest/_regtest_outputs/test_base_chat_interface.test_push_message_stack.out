[HumanMessage(content='Test 2', additional_kwargs={}, example=False), SystemMessage(content='Test 3', additional_kwargs={})]
[AIMessage(content='Test 1', additional_kwargs={}, example=False), HumanMessage(content='Test 2', additional_kwargs={}, example=False), SystemMessage(content='Test 3', additional_kwargs={})]
